client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoLegacyPreparedStatements.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.jdbc;
+
+import io.airlift.log.Logging;
+import io.trino.plugin.blackhole.BlackHolePlugin;
+import io.trino.plugin.tpch.TpchPlugin;
+import io.trino.server.testing.TestingTrinoServer;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.lang.reflect.Method;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Properties;
+import java.util.concurrent.ExecutorService;
+
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.lang.String.format;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static org.testng.Assert.assertEquals;
+
+public class TestTrinoLegacyPreparedStatements
```

Comment:
Let's use the tests in `TestJdbcPreparedStatement`?  We can parametrize `createConnection()` with a `useLegacyPreparedStatements` argument and then run all those tests in both modes.  And change some tests to use string parameters that contain apostrophes.

Reply:
Agreed. In particular, I believe this feature, as written, will break `PreparedStatement.getMetaData()` and `PreparedStatement.getParameterMetaData()`, which will be shown when `TestJdbcPreparedStatement` is run in this mode.

Reply:
The tests were moved to `TestJdbcPreparedStatement`

Reply:
That is not what was suggested.  We need to run all tests in that class in both modes, which will reveal some problems with the current implementation (see David's comment).  Also, I think that if you change some of the string arguments and statements in the current tests to contain apostrophes, you can probably get rid of the added tests that inspect the generated SQL (I think it's much easier to reason when a test checks that the result set matches the expectation than the statement issued).
```
    @DataProvider
    public Object[][] legacyPreparedStatementProvider()
    {
        return new Object[][] {{true}, {false}};
    }

    @Test(dataProvider = "legacyPreparedStatementProvider")
    public void testExecuteQuery(boolean legacyPreparedStatements)
            throws Exception
    {
        try (Connection connection = createConnection(legacyPreparedStatements);
                PreparedStatement statement = connection.prepareStatement("SELECT ?, ?")) {
            statement.setInt(1, 123);
            statement.setString(2, "hell'o");
...
```

Reply:
When I run with the parameter, all the tests succeed except for `testGetMetadata()` and `testGetParameterMetaData()` as both are looking for the prepared statement.

Reply:
The code was changed to avoid those two tests from breaking.